File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/circuitbreakers/RiskyVenueHandler.java
package uk.nhs.nhsx.circuitbreakers;

import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyResponseEvent;
import uk.nhs.nhsx.core.Environment;
import uk.nhs.nhsx.core.EnvironmentKeys;
import uk.nhs.nhsx.core.HttpResponses;
import uk.nhs.nhsx.core.SystemClock;
import uk.nhs.nhsx.core.auth.ApiName;
import uk.nhs.nhsx.core.auth.Authenticator;
import uk.nhs.nhsx.core.auth.ResponseSigner;
import uk.nhs.nhsx.core.aws.ssm.AwsSsmParameters;
import uk.nhs.nhsx.core.aws.ssm.ParameterName;
import uk.nhs.nhsx.core.aws.ssm.Parameters;
import uk.nhs.nhsx.core.events.CircuitBreakerVenueRequest;
import uk.nhs.nhsx.core.events.CircuitBreakerVenueResolution;
import uk.nhs.nhsx.core.events.Events;
import uk.nhs.nhsx.core.events.PrintingJsonEvents;
import uk.nhs.nhsx.core.routing.Routing;
import uk.nhs.nhsx.core.routing.RoutingHandler;

import java.time.Instant;
import java.util.function.Supplier;

import static uk.nhs.nhsx.circuitbreakers.CircuitBreakerService.startsWith;
import static uk.nhs.nhsx.core.StandardSigning.signResponseWithKeyGivenInSsm;
import static uk.nhs.nhsx.core.auth.StandardAuthentication.awsAuthentication;
import static uk.nhs.nhsx.core.routing.Routing.Method.GET;
import static uk.nhs.nhsx.core.routing.Routing.Method.POST;
import static uk.nhs.nhsx.core.routing.Routing.path;
import static uk.nhs.nhsx.core.routing.Routing.routes;
import static uk.nhs.nhsx.core.routing.StandardHandlers.authorisedBy;
import static uk.nhs.nhsx.core.routing.StandardHandlers.withSignedResponses;

public class RiskyVenueHandler extends RoutingHandler {

    private static final ParameterName initial = ParameterName.of("venue-notification-initial");
    private static final ParameterName poll = ParameterName.of("venue-notification-poll");

    private final Routing.Handler handler;

    @SuppressWarnings("unused")
    public RiskyVenueHandler() {
        this(Environment.fromSystem(), SystemClock.CLOCK, new PrintingJsonEvents(SystemClock.CLOCK));
    }

    public RiskyVenueHandler(Environment environment, Supplier<Instant> clock, Events events) {
        this(
            environment,
            awsAuthentication(ApiName.Mobile, events),
            signResponseWithKeyGivenInSsm(environment, clock, events),
            new AwsSsmParameters(),
            awsAuthentication(ApiName.Health, events),
            events
        );
    }

    public RiskyVenueHandler(Environment environment, Authenticator authenticator, ResponseSigner signer, Parameters parameters, Authenticator healthAuthenticator, Events events) {
        this(environment,
            authenticator,
            signer,
            new CircuitBreakerService(
                parameters.ofEnum(initial.withPrefix(environment.access.required(EnvironmentKeys.SSM_CIRCUIT_BREAKER_BASE_NAME)), ApprovalStatus.class, ApprovalStatus.PENDING),
                parameters.ofEnum(poll.withPrefix(environment.access.required(EnvironmentKeys.SSM_CIRCUIT_BREAKER_BASE_NAME)), ApprovalStatus.class, ApprovalStatus.PENDING)
            ),
            events,
            healthAuthenticator);
    }

    public RiskyVenueHandler(Environment environment, Authenticator authenticator, ResponseSigner signer, CircuitBreakerService circuitBreakerService, Events events, Authenticator healthAuthenticator) {
        handler = withSignedResponses(
            events,
            environment,
            signer,
            routes(
                authorisedBy(authenticator,
                    path(POST, startsWith("/circuit-breaker/venue/request"),
                        r -> {
                            events.emit(getClass(), new CircuitBreakerVenueRequest());
                            return mapResultToResponse(circuitBreakerService.getApprovalToken());
                        }
                    )
                ),
                authorisedBy(authenticator,
                    path(GET, startsWith("/circuit-breaker/venue/resolution"),
                        r -> {
                            events.emit(getClass(), new CircuitBreakerVenueResolution());
                            return mapResultToResponse(circuitBreakerService.getResolution(r.getPath()));
                        }
                    )
                ),
                authorisedBy(healthAuthenticator,
                    path(POST, "/circuit-breaker/venue/health", r ->
                        HttpResponses.ok()
                    )
                )
            )
        );
    }

    private APIGatewayProxyResponseEvent mapResultToResponse(CircuitBreakerResult result) {
        if (result.type == CircuitBreakerResult.ResultType.ValidationError) {
            return HttpResponses.unprocessableEntity("validation error: Content type is not text/json");
        }
        if (result.type == CircuitBreakerResult.ResultType.MissingPollingTokenError) {
            return HttpResponses.unprocessableEntity("missing polling token error: Request was submitted without a polling token");
        }
        return HttpResponses.ok(result.responseBody);
    }

    @Override
    public Routing.Handler handler() {
        return handler;
    }
}

File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/circuitbreakers/RiskyVenueHandler.kt
package uk.nhs.nhsx.circuitbreakers

import com.amazonaws.services.kms.AWSKMSClientBuilder
import com.amazonaws.services.lambda.runtime.Context
import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyRequestEvent
import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyResponseEvent
import uk.nhs.nhsx.core.Clock
import uk.nhs.nhsx.core.Environment
import uk.nhs.nhsx.core.EnvironmentKeys
import uk.nhs.nhsx.core.HttpResponses
import uk.nhs.nhsx.core.StandardSigningFactory
import uk.nhs.nhsx.core.SystemClock
import uk.nhs.nhsx.core.auth.ApiName
import uk.nhs.nhsx.core.auth.Authenticator
import uk.nhs.nhsx.core.auth.ResponseSigner
import uk.nhs.nhsx.core.auth.StandardAuthentication.awsAuthentication
import uk.nhs.nhsx.core.aws.ssm.AwsSsmParameters
import uk.nhs.nhsx.core.aws.ssm.ParameterName.Companion.of
import uk.nhs.nhsx.core.aws.ssm.Parameters
import uk.nhs.nhsx.core.aws.ssm.ofEnum
import uk.nhs.nhsx.core.events.CircuitBreakerVenueRequest
import uk.nhs.nhsx.core.events.CircuitBreakerVenueResolution
import uk.nhs.nhsx.core.events.Events
import uk.nhs.nhsx.core.events.PrintingJsonEvents
import uk.nhs.nhsx.core.handler.ApiGatewayHandler
import uk.nhs.nhsx.core.routing.Routing
import uk.nhs.nhsx.core.routing.Routing.path
import uk.nhs.nhsx.core.routing.Routing.routes
import uk.nhs.nhsx.core.handler.RoutingHandler
import uk.nhs.nhsx.core.routing.authorisedBy
import uk.nhs.nhsx.core.routing.withSignedResponses

class RiskyVenueHandler @JvmOverloads constructor(
    environment: Environment = Environment.fromSystem(),
    clock: Clock = SystemClock.CLOCK,
    events: Events = PrintingJsonEvents(clock),
    authenticator: Authenticator = awsAuthentication(ApiName.Mobile, events),
    parameters: Parameters = AwsSsmParameters(),
    signer: ResponseSigner = StandardSigningFactory(
        clock,
        parameters,
        AWSKMSClientBuilder.defaultClient()
    ).signResponseWithKeyGivenInSsm(environment, events),
    circuitBreakerService: CircuitBreakerService = CircuitBreakerService(
        parameters.ofEnum(
            initial.withPrefix(environment.access.required(EnvironmentKeys.SSM_CIRCUIT_BREAKER_BASE_NAME)),
            ApprovalStatus::class.java,
            ApprovalStatus.PENDING
        ),
        parameters.ofEnum(
            poll.withPrefix(environment.access.required(EnvironmentKeys.SSM_CIRCUIT_BREAKER_BASE_NAME)),
            ApprovalStatus::class.java,
            ApprovalStatus.PENDING
        )
    ),
    healthAuthenticator: Authenticator = awsAuthentication(ApiName.Health, events)
) : RoutingHandler() {
    private val handler: ApiGatewayHandler

    private fun mapResultToResponse(result: CircuitBreakerResult): APIGatewayProxyResponseEvent {
        if (result.type == CircuitBreakerResult.ResultType.ValidationError) {
            return HttpResponses.unprocessableEntity("validation error: Content type is not text/json")
        }
        return if (result.type == CircuitBreakerResult.ResultType.MissingPollingTokenError) {
            HttpResponses.unprocessableEntity("missing polling token error: Request was submitted without a polling token")
        } else HttpResponses.ok(result.responseBody)
    }

    override fun handler(): ApiGatewayHandler {
        return handler
    }

    companion object {
        private val initial = of("venue-notification-initial")
        private val poll = of("venue-notification-poll")
    }

    init {
        handler = withSignedResponses(
            events,
            environment,
            signer,
            routes(
                authorisedBy(
                    authenticator,
                    path(
                        Routing.Method.POST, CircuitBreakerService.startsWith("/circuit-breaker/venue/request"),
                        ApiGatewayHandler { _, _ ->
                            events(CircuitBreakerVenueRequest())
                            mapResultToResponse(circuitBreakerService.getApprovalToken())
                        }
                    )
                ),
                authorisedBy(
                    authenticator,
                    path(
                        Routing.Method.GET, CircuitBreakerService.startsWith("/circuit-breaker/venue/resolution"),
                        ApiGatewayHandler { r, _ ->
                            events(CircuitBreakerVenueResolution())
                            mapResultToResponse(circuitBreakerService.getResolution(r.path))
                        }
                    )
                ),
                authorisedBy(
                    healthAuthenticator,
                    path(
                        Routing.Method.POST,
                        "/circuit-breaker/venue/health",
                        ApiGatewayHandler { _: APIGatewayProxyRequestEvent, _: Context -> HttpResponses.ok() }
                    )
                )
            )
        )
    }
}
